AssertionError(msg="`Union{}` typed field should be strictly undefined")

The Oscar.jl tests seem to be running into this assertion now after #57541 tried to fix https://github.com/oscar-system/Oscar.jl/issues/4648: 
```
 From worker 4:	Internal error: during type inference of
      From worker 4:	(::Type{Oscar.IntersectionTheory.AbstractVarietyMap{V1, V2} where V2<:Oscar.IntersectionTheory.AbstractVarietyT where V1<:Oscar.IntersectionTheory.AbstractVarietyT})(Oscar.IntersectionTheory.AbstractVariety, Oscar.IntersectionTheory.AbstractVariety, Array{Oscar.MPolyQuoRingElem{Oscar.MPolyDecRingElem{Nemo.QQFieldElem, Nemo.QQMPolyRingElem}}, 1}, AbstractAlgebra.Generic.FunctionalMap{Oscar.MPolyQuoRing{Oscar.MPolyDecRingElem{Nemo.QQFieldElem, Nemo.QQMPolyRingElem}}, Oscar.MPolyDecRing{Nemo.QQFieldElem, Nemo.QQMPolyRing}})
      From worker 4:	Encountered unexpected error in runtime:
      From worker 4:	AssertionError(msg="`Union{}` typed field should be strictly undefined")
      From worker 4:	PartialStruct at ./coreir.jl:62
```
https://github.com/oscar-system/Oscar.jl/actions/runs/13611555768/job/38059503585#step:11:677

_Originally posted by @benlorenz in https://github.com/JuliaLang/julia/pull/57541#discussion_r1976640820_

cc @benlorenz @fingolfin 

ping @aviatesk @serenity4 @topolarity as you guys worked on #57541

The following case seems to still be unsupported:
```julia
julia> struct A{T}; x::Any; y::T; A{T}(x) where {T} = new{T}(x); end

julia> Core.PartialStruct(A{Union{}}, Union{Nothing,Bool}[false, nothing], Any[Int, Union{}])
ERROR: AssertionError: `Union{}` typed field should be strictly undefined
Stacktrace:
 [1] Core.PartialStruct(typ::Type, undefs::Vector{Union{Nothing, Bool}}, fields::Vector{Any})
   @ Base ./coreir.jl:62
```

Perhaps we should adjust the check to also allow `undef === nothing`, instead of requiring undef to be `true` for `Union{}` fields.

While a field with type `Union{}` implies `undef === true`, it seems a bit too much to me to require all callsites to check for `Union{}` in every field type to determine whether the corresponding `undef` entry should be set to `true` before constructing a `PartialStruct`.

Here it goes:
```julia
julia> using Oscar, Oscar.IntersectionTheory

julia> code_typed(degeneracy_locus, (Oscar.IntersectionTheory.AbstractBundle{Oscar.IntersectionTheory.AbstractVariety}, Oscar.IntersectionTheory.AbstractBundle{Oscar.IntersectionTheory.AbstractVariety}, Int64))

ERROR: AssertionError: `Union{}` typed field should be strictly undefined
Stacktrace:
  [1] Core.PartialStruct(typ::Type, undefs::Vector{Union{Nothing, Bool}}, fields::Vector{Any})
    @ Base ./coreir.jl:62
  [2] Core.PartialStruct(::Compiler.InferenceLattice{Compiler.ConditionalsLattice{…}}, typ::Any, undefs::Vector{Union{…}}, fields::Vector{Any})
    @ Compiler ./../usr/share/julia/Compiler/src/typelattice.jl:765
  [3] form_partially_defined_struct(𝕃ᵢ::Compiler.InferenceLattice{Compiler.ConditionalsLattice{Compiler.PartialsLattice{Compiler.ConstsLattice}}}, obj::Any, name::Any)
    @ Compiler ~/.julia/juliaup/julia-nightly/share/julia/Compiler/src/abstractinterpretation.jl:2162
  [4] abstract_call_builtin(interp::Compiler.NativeInterpreter, f::Core.Builtin, ::Compiler.ArgInfo, sv::Compiler.InferenceState)
    @ Compiler ~/.julia/juliaup/julia-nightly/share/julia/Compiler/src/abstractinterpretation.jl:2118
  [5] abstract_call_known(interp::Compiler.NativeInterpreter, f::Any, arginfo::Compiler.ArgInfo, si::Compiler.StmtInfo, sv::Compiler.InferenceState, max_methods::Int64)
    @ Compiler ~/.julia/juliaup/julia-nightly/share/julia/Compiler/src/abstractinterpretation.jl:2664
  [...] (truncated)
```

It errors when constructing a `PartialStruct` with
```julia
objt0 = Oscar.MPolyAnyMap{<:AbstractAlgebra.MPolyRing{T}, Union{}, Nothing} where T<:AbstractAlgebra.FieldElem
undefs = Union{Nothing, Bool}[0, 0, 0, 0, nothing, nothing, 0]
fields = Any[AbstractAlgebra.MPolyRing{T} where T<:AbstractAlgebra.FieldElem, Union{}, Nothing, Vector, Any, Vector{Int64}, Dict{Symbol, Any}]
```

where the `Oscar.MPolyAnyMap` type is defined here: https://github.com/oscar-system/Oscar.jl/blob/43891614c618991a89f198068f4d514d89a3e3d6/src/Rings/MPolyMap/Types.jl#L4-L42

We can see it being instantiated with `C = Union{}` type parameter, and because the inner constructor calls `new` with 4 values [here](https://github.com/oscar-system/Oscar.jl/blob/43891614c618991a89f198068f4d514d89a3e3d6/src/Rings/MPolyMap/Types.jl#L30), `datatype_min_ninitialized` returns 4 and therefore `partialstruct_init_undefs` fills in the corresponding `undefs` entries with `false`.

Therefore, the fix I suggested would not even be enough. Even if we may wonder why it occurs, isn't that technically a valid Julia type that we may see in inference? (even if it would never be constructed at runtime)

I guess a possible fix would be to look at `fields` in `partialstruct_init_undefs` and set the corresponding entry to `true` for `Union{}` fields. In theory it might be enough so long as any construction of `PartialStruct` uses that function to define `undefs`.

--- a/Compiler/src/abstractinterpretation.jl
+++ b/Compiler/src/abstractinterpretation.jl
@@ -2156,7 +2156,11 @@ function form_partially_defined_struct(𝕃ᵢ::AbstractLattice, @nospecialize(o
     if fields[fldidx] === Union{}
         return nothing # `Union{}` field never transitions to be defined
     end
-    undefs = partialstruct_init_undefs(objt, fldcnt)
+    undefs = partialstruct_init_undefs(objt, fields)
+    if undefs === nothing
+        # this object never exists at runtime, avoid creating unprofitable `PartialStruct`
+        return nothing
+    end
     undefs[fldidx] = false
     return PartialStruct(𝕃ᵢ, objt0, undefs, fields)
 end

--- a/Compiler/src/tfuncs.jl
+++ b/Compiler/src/tfuncs.jl
@@ -1995,7 +1995,7 @@ function tuple_tfunc(𝕃::AbstractLattice, argtypes::Vector{Any})
     typ = Tuple{params...}
     # replace a singleton type with its equivalent Const object
     issingletontype(typ) && return Const(typ.instance)
-    return anyinfo ? PartialStruct(𝕃, typ, partialstruct_init_undefs(typ, argtypes), argtypes) : typ
+    return anyinfo ? PartialStruct(𝕃, typ, partialstruct_init_undefs(typ, argtypes)::Vector, argtypes) : typ
 end
 
 @nospecs function memorynew_tfunc(𝕃::AbstractLattice, memtype, memlen)

--- a/Compiler/src/typelattice.jl
+++ b/Compiler/src/typelattice.jl
@@ -755,7 +755,9 @@ end
 
 # Legacy constructor
 function Core.PartialStruct(𝕃::AbstractLattice, @nospecialize(typ), fields::Vector{Any})
-    return PartialStruct(𝕃, typ, partialstruct_init_undefs(typ, fields), fields)
+    undefs = partialstruct_init_undefs(typ, fields)
+    undefs === nothing && error("This object never exists at runtime")
+    return PartialStruct(𝕃, typ, undefs, fields)
 end
 
 function Core.PartialStruct(::AbstractLattice, @nospecialize(typ), undefs::Vector{Union{Nothing,Bool}}, fields::Vector{Any})

--- a/Compiler/test/inference.jl
+++ b/Compiler/test/inference.jl
@@ -6382,4 +6382,15 @@ g57292(xs::String...) = getfield(("abc",), 1, :not_atomic, xs...)
 @test Base.infer_return_type(f57292) == String
 @test Base.infer_return_type(g57292) == String
 
+mutable struct Issue57673{C<:Union{Int,Float64}}
+    c::C
+    d
+    Issue57673(c::C, d) where C = new{C}(c, d)
+    Issue57673(c::C) where C = new{C}(c)
+end
+@test Base.infer_return_type((Issue57673,)) do a::Issue57673{<:String}
+    setfield!(a, :d, nothing)
+    a
+end === Union{} # `setfield!` tfunc should be able to figure out this object is runtime invalid
+
 end # module inference

--- a/base/coreir.jl
+++ b/base/coreir.jl
@@ -67,14 +67,27 @@ end
 
 # Legacy constructor
 function Core.PartialStruct(@nospecialize(typ), fields::Vector{Any})
-    return Core.PartialStruct(typ, partialstruct_init_undefs(typ, fields), fields)
+    undefs = partialstruct_init_undefs(typ, fields)
+    undefs === nothing && error("This object never exists at runtime")
+    return Core.PartialStruct(typ, undefs, fields)
 end
 
-partialstruct_init_undefs(@nospecialize(typ), fields::Vector{Any}) = partialstruct_init_undefs(typ, length(fields))
-function partialstruct_init_undefs(@nospecialize(typ), n::Int)
-    undefs = Union{Nothing,Bool}[nothing for _ in 1:n]
-    for i in 1:min(datatype_min_ninitialized(typ), n)
+function partialstruct_init_undefs(@nospecialize(typ), fields::Vector{Any})
+    nf = length(fields)
+    minf = datatype_min_ninitialized(typ)
+    for i = 1:minf
+        if fields[i] === Union{}
+            return nothing # disallow runtime-invalid `PartialStruct`
+        end
+    end
+    undefs = Union{Nothing,Bool}[nothing for _ in 1:nf]
+    for i in 1:minf
         undefs[i] = false
+    end
+    for i = minf+1:nf
+        if fields[i] === Union{}
+            undefs[i] = true
+        end
     end
     return undefs
 end

